# Project: Hacker News Data analysis

## Table of Contents
* Introduction
* Data Wrangling
* Exploratory data analysis
* Conclusion

# Introduction
> Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result. I would be asking the following questions from this dataset:
> 
> 1. Do Ask HN or Show HN receive more comments on average?
> 2. Do posts created at a certain time receive more comments on average?
>
> Users submit Ask HN posts to ask the Hacker News community a specific question.
> Users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

# Data Wrangling

## Gather 
Let's start by importing the libraries we need

In [1]:
from csv import reader

## Loading data

In [2]:
# Use the python built in function open() to open the hacker news data
opened_file = open('hacker_news.csv')

# Use csv.reader() to parse the data from the opened file
read_file = reader(opened_file)

# Use list() to convert the read file into a list of lists format
hn1 = list(read_file)

#Close the opened file
opened_file.close()

# Remove the first row of the data, which contains the column names
hn = hn1[1:]

## Assess data

In [3]:
# Print number of rows and column
print('Number of rows:', len(hn))
print('Number of columns:', len(hn[0]))

Number of rows: 20100
Number of columns: 7


In [4]:
# Print column names
hn1[0]

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

Below are descriptions of the columns:

* id: The unique identifier from Hacker News for the post
* title: The title of the post
* url: The URL that the posts links to, if it the post has a URL
* num_points: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
* num_comments: The number of comments that were made on the post
* author: The username of the person who submitted the post
* created_at: The date and time at which the post was submitte

In [5]:
#Create a function named explore_data that can be repeatedly used to print rows in a readable way
def explore_data(dataset, start, end):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row
        
explore_data(hn, 0, 10) #Prints the first 10 rows

['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']


['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']


['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']


['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']


['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']


['10482257', 'Title II kills investment? Comcast and other ISPs are now spending more', '

In [6]:
# lets store our column name into a variable headers
headers = hn1[0]
headers

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

## Cleaning Data

### Issue
I'm only concerned with post titles beginning with **Ask HN** or **Show HN**

### Define
I'll create new lists of lists containing just the data for those titles using the string method **startswith**

### Code

In [7]:
# Create three empty lists called ask_posts, show_posts, and other_posts
ask_posts = []
show_posts = []
other_posts = []
for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

### Test

In [8]:
len(ask_posts)

1744

In [9]:
len(show_posts)

1162

In [10]:
len(other_posts)

17194

In [11]:
1744+1162+17194

20100

This shows thst our cleaning worked and we have in total;
* 1744 ask hn posts 
* 1162 show hn posts

In [12]:
# first five rows in ask_posts
ask_posts[0:5]

[['12296411',
  'Ask HN: How to improve my personal website?',
  '',
  '2',
  '6',
  'ahmedbaracat',
  '8/16/2016 9:55'],
 ['10610020',
  'Ask HN: Am I the only one outraged by Twitter shutting down share counts?',
  '',
  '28',
  '29',
  'tkfx',
  '11/22/2015 13:43'],
 ['11610310',
  'Ask HN: Aby recent changes to CSS that broke mobile?',
  '',
  '1',
  '1',
  'polskibus',
  '5/2/2016 10:14'],
 ['12210105',
  'Ask HN: Looking for Employee #3 How do I do it?',
  '',
  '1',
  '3',
  'sph130',
  '8/2/2016 14:20'],
 ['10394168',
  'Ask HN: Someone offered to buy my browser extension from me. What now?',
  '',
  '28',
  '17',
  'roykolak',
  '10/15/2015 16:38']]

In [13]:
#First 5 rows in show_posts
show_posts[0:5]

[['10627194',
  'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform',
  'https://iot.seeed.cc',
  '26',
  '22',
  'kfihihc',
  '11/25/2015 14:03'],
 ['10646440',
  'Show HN: Something pointless I made',
  'http://dn.ht/picklecat/',
  '747',
  '102',
  'dhotson',
  '11/29/2015 22:46'],
 ['11590768',
  'Show HN: Shanhu.io, a programming playground powered by e8vm',
  'https://shanhu.io',
  '1',
  '1',
  'h8liu',
  '4/28/2016 18:05'],
 ['12178806',
  'Show HN: Webscope  Easy way for web developers to communicate with Clients',
  'http://webscopeapp.com',
  '3',
  '3',
  'fastbrick',
  '7/28/2016 7:11'],
 ['10872799',
  'Show HN: GeoScreenshot  Easily test Geo-IP based web pages',
  'https://www.geoscreenshot.com/',
  '1',
  '9',
  'kpsychwave',
  '1/9/2016 20:45']]

# Exploratory data analysis

## Research Question 1 (Do Ask HN or Show HN receive more comments on average?)

### Compute the average number of comments on ask posts

In [14]:
# Find the total number of comments in ask posts and assign it to total_ask_comments
total_ask_comments = 0
for row in ask_posts:
    num_comments = row[4]
    num_comments = int(num_comments)
    total_ask_comments += num_comments
    
#Compute the average number of comments on ask posts and assign it to avg_ask_comments
avg_ask_comments = total_ask_comments/1744
print(avg_ask_comments)

14.038417431192661


### Compute the average number of comments on ask posts

In [15]:
# Find the total number of comments in show posts and assign it to total_show_comments
total_show_comments = 0
for row in show_posts:
    num_comments = row[4]
    num_comments = int(num_comments)
    total_show_comments += num_comments
    
#Compute the average number of comments on show posts and assign it to avg_show_comments
avg_show_comments = total_show_comments/1162
print(avg_show_comments)

10.31669535283993


On average, ask posts in our sample receive approximately 14 comments, whereas show posts receive approximately 10. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

## Research Question 2 (Do posts created at a certain time receive more comments on average?)- Since ask posts are more likely to receive comments, we'll focus our analysis just on these posts

We'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:
1. Calculate the amount of ask posts created in each hour of the day, along with the number of comments received.
2. Calculate the average number of comments ask posts receive by hour created.

### Finding the Amount of Ask Posts and Comments by Hour Created

In [16]:
#find out the date format used in created_at column from a random row in ask_posts
ask_posts[10][6]

'9/19/2015 17:04'

date format used is "%m/%d/%Y %H:%M"

In [17]:
# import datetime module as dt
import datetime as dt

result_list = []
for row in ask_posts:
    created_at = row[6]
    num_comments = int(row[4])
    result_list.append([created_at, num_comments])
    
counts_by_hour = {}
comments_by_hour = {}

date_format = "%m/%d/%Y %H:%M"
for row in result_list:
    date = row[0]
    comments = row[1]
    date_dt = dt.datetime.strptime(date, date_format)
    date_str = date_dt.strftime("%H")
    if date_str in counts_by_hour:
        counts_by_hour[date_str] += 1
        comments_by_hour[date_str] += comments
    else:
        counts_by_hour[date_str] = 1
        comments_by_hour[date_str] = comments

In [18]:
comments_by_hour

{'09': 251,
 '13': 1253,
 '10': 793,
 '14': 1416,
 '16': 1814,
 '23': 543,
 '12': 687,
 '17': 1146,
 '15': 4477,
 '21': 1745,
 '20': 1722,
 '02': 1381,
 '18': 1439,
 '03': 421,
 '05': 464,
 '19': 1188,
 '01': 683,
 '22': 479,
 '08': 492,
 '04': 337,
 '00': 447,
 '06': 397,
 '07': 267,
 '11': 641}

In [19]:
counts_by_hour

{'09': 45,
 '13': 85,
 '10': 59,
 '14': 107,
 '16': 108,
 '23': 68,
 '12': 73,
 '17': 100,
 '15': 116,
 '21': 109,
 '20': 80,
 '02': 58,
 '18': 109,
 '03': 54,
 '05': 46,
 '19': 110,
 '01': 60,
 '22': 71,
 '08': 48,
 '04': 47,
 '00': 55,
 '06': 44,
 '07': 34,
 '11': 58}

### Calculating the average number of comments ask posts receive by hour created

Create a list of lists containing the hours during which posts were created and the average number of comments those posts received.

In [20]:
# Calculate the average amount of comments `Ask HN` posts created at each hour of the day receive.
avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])

avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

Let's finish by sorting the list of lists and printing the five highest values in a format that's easier to read.

In [21]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])

print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse = True)

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


In [22]:
sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [23]:
print("Top 5 Hours for Ask Posts Comments.")
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for Ask Posts Comments.
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


From the results above, The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post

# Conclusion

In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).